src/thor/costmatrix.cc

```diff
@@ -948,82 +956,93 @@ void CostMatrix::SetTargets(baldr::GraphReader& graphreader,
   }
 }
 
-// Form the path from the adjacency list.
-// TODO: move this function to PathInfo header or so, where both bidir A* and CostMatrix
-// can see it
-void CostMatrix::RecostPaths(GraphReader& graphreader,
-                             google::protobuf::RepeatedPtrField<valhalla::Location>& sources,
-                             google::protobuf::RepeatedPtrField<valhalla::Location>& targets,
-                             const std::vector<baldr::TimeInfo>& time_infos,
-                             bool invariant) {
-  uint32_t idx = 0;
-  for (auto best_connection = best_connection_.begin(); best_connection != best_connection_.end();
-       ++best_connection, ++idx) {
-    // no need to look at source == target or missing connectivity
-    if (best_connection->cost.secs == 0.f || best_connection->distance == kMaxCost) {
-      continue;
-    }
+// Form the path from the edfge labels and optionally return the shape
+std::string CostMatrix::RecostFormPath(GraphReader& graphreader,
+                                       BestCandidate& connection,
+                                       const valhalla::Location& source,
+                                       const valhalla::Location& target,
+                                       const uint32_t source_idx,
+                                       const uint32_t target_idx,
+                                       const baldr::TimeInfo& time_info,
+                                       const bool invariant,
+                                       const ShapeFormat shape_format) {
+  // no need to look at source == target or missing connectivity
+  if ((!has_time_ && shape_format == no_shape) || connection.cost.secs == 0.f ||
+      connection.distance == kMaxCost) {
+    return "";
+  }
 
-    uint32_t source_idx = idx / static_cast<uint32_t>(targets.size());
-    uint32_t target_idx = idx % static_cast<uint32_t>(targets.size());
```

Comment:
this is a terrible diff: I changed no logic here. just removed the for loop (as the caller is doing that)

Reply:
:pray: